### PR TITLE
Implement patch checking during pre-commit hook

### LINKED ISSUE #28: Allow readme syncing to be disabled
There are cases where one might want the other stuff, but would like to turn of the readme syncing.

I think it would be generally useful to add support in the precommit file for an additional configuration file to be loaded, so that it could be configured/extended without hacking it. I guess that might be for another issue, though.

### LINKED ISSUE #30: Introduce .ci-config.sh to override Travis and pre-commit defaults
There should be a `.ci-config.sh` (or something) which exports various Bash variables which can be read-in by `.travis.yml` and `bin/pre-commit`.These variables can control things like:

- [x] Allow readme syncing to be disabled (#28)
- [x] Where to fetch PHP_CodeSniffer from (also should grab from GitHub instead of Pear), and which branch or tag to use.
- [x] Where to fetch WordPress Coding Standards from, and which branch/tag/commit to use.
- [x] Path to `phpunit.xml`
- [x] Path to PHPCS ruleset XML file.
- [ ] ~~URL for Travis build button (for `generate-markdown-readme`), instead of pulling from the `git remote`~~

Others?

### LINKED ISSUE #40: Reduce redundancy between pre-commit and .travis.yml files
Consolidate default environment variables into a common script included by both Travis and `pre-commit` hook. Reduce redundancy.

### LINKED ISSUE #49: Prevent invoking YUI Compressor if there are no JS files to check

### LINKED ISSUE #77: Integrate phpcs-patch into Travis and pre-commit hook
We need to facilitate only reporting PHP_CodeSniffer issues for lines that are part of changes in the pull request (on Travis) or as part of the current commit (Travis or `pre-commit` hook).

This depends on https://github.com/xwp/PHP_CodeSniffer/pull/5

When running on Travis, we need to look at the `TRAVIS_COMMIT_RANGE` environment variable. For `pre-commit`, we obviously just look at the stage.

The work here is needed for integrating PHP_CodeSniffer into the WordPress Core build tools ([#30153](https://core.trac.wordpress.org/ticket/30153)).

### LINKED ISSUE #100: Only check patches in pre-commit hook
Implement the same thing that was done for Travis https://github.com/xwp/wp-dev-lib/pull/95 to also do on the `pre-commit` hook. We're already limiting checks to modified files, but we should limit the errors reported to just the lines changed in the files.

### LINKED ISSUE #126: PHPCS and WPCS should be downloaded in pre-commit if not already installed

### PR DESCRIPTION
**Pull request by westonruter**
* Patch checking is aware of the Git index, so only staged changes will be checked.
* Consolidates pre-commit and Travis CI logic into common library.
* Moves Travis before_script logic into install script.

Fixes #126
Fixes #100 
Fixes #77
Fixes #40
Fixes #49
Fixes #30 
Fixes #28

### CODE REVIEW

**Comment by valendesigns** [comment]
Looks good! We'll need to test this PR out with some repos to verify backwards compatibility.

**Comment by westonruter** [comment]
Thanks.

Back compat isn't so important as it working. The .travis.yml will have to
be updated to refer to the install script. Projects using dev-lib  won't
break if this gets merged because the submodule will be pegging it as a
compatible state.

I have tested this in the pre-commit hook but I need to run through the
tests for Travis still.

This should be a huge improvement to ensuring quality across repos, as
PHPCS is downloaded if not already installed and it checks patches instead
of the full codebases. This has been a long time in coming!

**Comment by westonruter** [comment]
I'm also wary of Windows compatibility, but I'm not going to worry too much
about that :-)

**Comment by westonruter** [comment]
I went through and tested the scenarios in Travis: https://github.com/xwp/wp-customize-widgets-plus/pull/41

And then I made the necessary fixes here.

The required changes to support the new dev-lib can be seen here: https://github.com/xwp/wp-customize-widgets-plus/issues/42

@valendesigns See anything else I missed? If not, please merge.

File: check-diff.sh [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+		else
+			PROJECT_TYPE=unknown
+		fi
+	fi
+
+	if [ ! -z "$LIMIT_TRAVIS_PR_CHECK_SCOPE" ]; then
+		echo "LIMIT_TRAVIS_PR_CHECK_SCOPE is obsolete; use CHECK_SCOPE env var instead" 1>&2
+		return 1
+	fi
+	CHECK_SCOPE=${CHECK_SCOPE:-changed-files} # 'all', 'changed-files', 'patches'
```
Comment (westonruter) [inline check-diff.sh]: Maybe `patches` should be the default?

**Comment by valendesigns** [comment]
I made the change you wanted to default to `patches` instead of `changed-files` and then updated the Authoring plugin. It appeared to go well so I'll merge tomorrow. I do want to test it on https://github.com/xwp/wp-customize-snapshots or https://github.com/xwp/wp-customize-concurrency to be sure nothing else is needed before merging though. :smile:

**Pull request merged**